vortex-array/src/array/constant/compute.rs

```diff
@@ -59,6 +68,45 @@ impl SearchSortedFn for ConstantArray {
     }
 }
 
+impl AndFn for ConstantArray {
+    fn and(&self, array: &Array) -> VortexResult<Array> {
+        constant_array_bool_impl(self, array, |(l, r)| l & r)
+    }
+}
+
+impl OrFn for ConstantArray {
+    fn or(&self, array: &Array) -> VortexResult<Array> {
+        constant_array_bool_impl(self, array, |(l, r)| l | r)
+    }
+}
+
+fn constant_array_bool_impl(
+    constant_array: &ConstantArray,
+    other: &Array,
+    bool_op: impl Fn((bool, bool)) -> bool,
+) -> VortexResult<Array> {
+    if constant_array.dtype().is_boolean()
+        && other.dtype().is_boolean()
+        && constant_array.len() == other.len()
+    {
+        if let Ok(array) = ConstantArray::try_from(other.clone()) {
+            let lhs = constant_array.scalar().value().as_bool()?;
+            let rhs = array.scalar().value().as_bool()?;
+
+            let scalar = match lhs.zip(rhs).map(bool_op) {
+                Some(b) => Scalar::bool(b, Nullability::Nullable),
+                None => Scalar::null(constant_array.dtype().as_nullable()),
+            };
+
+            Ok(ConstantArray::new(scalar, constant_array.len()).into_array())
+        } else {
+            AndFn::and(&constant_array.clone().into_bool()?, other)
```

Comment:
Either this fallback will fail (into_bool when it's not?), or into_bool does an expensive expansion of the constant.

I think the best thing is to have the convention that constant goes on the RHS, so we should fallback to `and(other, constant_array)` to allow the RHS encoding a chance to be efficient, e.g. if it's run-end it only needs to run over the values.

Reply:
I can't actually fallback into `AndFn` here because this might also be an `or`, I'm trying to think of a better way of generalizing boolean ops here (I do agree with trying to let `rhs` to give us a more efficient implementation)